bootloader/bootloader.go

```diff
@@ -310,3 +311,56 @@ func removeKernelAssetsFromBootDir(bootDir string, s snap.PlaceInfo) error {
 
 	return nil
 }
+
+// kernelCommandLineSplit tries to split the string into a list of elements that
+// would be passed by the bootloader as the kernel command arguments.
+//
+// See https://www.kernel.org/doc/html/latest/admin-guide/kernel-parameters.html for details.
+func kernelCommandLineSplit(s string) (out []string, err error) {
+	var quoting bool
+	var b bytes.Buffer
+	var rs = []rune(s)
+	var last = len(rs) - 1
+	// arguments are:
+	// - arg
+	// - arg=value, where value can be any string, spaces are preserve when quoting ".."
+	for idx, r := range rs {
+		maybeSplit := false
+		switch r {
+		case '"':
+			if !quoting {
+				if b.Len() < 2 || idx == 0 || (idx > 0 && rs[idx-1] != '=') {
+					// either:
+					// - the whole input starts with "
+					// - preceding character is not =
+					// - there's no at least `a=` collected so far
+					return nil, fmt.Errorf("unexpected quoting")
+				}
+				quoting = true
+			} else {
+				quoting = false
```

Comment:
If we find a closing quote we also expect the end of that parameter, but validating the parameter format seems to be beyond our goal here. As long as the violation doesn't cause quote imbalance parameter splitting is not affected (and things like `a="b"="c"` or `a="b""c"` will pass). Just mentioning it to confirm that this is what's intended here.

Reply:
Heh, hoped to avoid adding extra complexity, but I think we should at least validate that the formatting makes some sort of sense. I've rewritten the splitting into a mini state machine, so it should be correct now.

Reply:
Ah nice, I think a state machine is safer for this kind of parsing. (Also a correction: with the previous parser my second example above would be detected, but `a=="b"` would not.)